---
title: "Supplementary Notebook 1 (Figure 5)"
output: html_document
date: "2025-03-03"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(snpR); library(PCAssess); library(ggplot2); library(data.table)
```

# Import and prepare
First we need to read in and filter our raw data. We'll then save our data after calculating $F_{ST}$ and taking the top 5% in case we want to do anything else with it later. We'll use the R package (snpR)[https://github.com/hemstrow/snpR] to do the filtering and calculate $F_{ST}$. snpR uses the same methods as PCAssess for this, so it will be comparable. 

Note that we took the 10,000 random SNPs to use here, since the original dataset is large and thus slow to run.
```{r, eval=FALSE}
# subet and filter
PSd <- readRDS("../data/even_filtered_maf05.RDS")
PSd <- PSd[origin = "natural"]
PSd <- filter_snps(PSd, maf = 0.05, maf_facets = "location", min_ind = .75, min_loci = .75)
set.seed(12321)
PSd <- PSd[sample(nrow(PSd), 10000, FALSE),]

# get high Fst loci for later
PSd <- calc_global_fst(PSd, "location") # FST; location~origin
saveRDS(PSd, "../data/PS_all.RDS")
PSd_high <- as.data.table(get.snpR.stats(PSd, "location", "fst")$pairwise)
PSd_high <- unlist(PSd_high[fst >= quantile(fst, .95, na.rm = TRUE),.(.snp.id)])
PSd_high <- PSd[which(snp.meta(PSd)$.snp.id %in% PSd_high),]
saveRDS(PSd_high, "../ata/PS_high.RDS")

```

We only need to run the above once--if we wish to re-run anything, we can pull back in the saved datasets.
```{r}
PSd <- readRDS("../data/PS_all.RDS")
PSd_high <- readRDS("../data/PS_high.RDS")
```


# Run PCAssess
Next, we'll use PCAssess to run 100 permutations in which we randomly shuffle population IDs, calculate $F_{ST}$, and construct PCAs for the top 5% of loci. PCAssess will then calculate a p-value for us based on the changes in clustering observed in the top loci for both our real population IDs and the permutations.
```{r}
# grab population info and format for PCAssess
facet <- sample.meta(PSd)$location
sn <- format_snps(PSd, "sn")
sn <- sn[,-c(1:3)]
head(sn)[,1:6]
```

Note that the import format here is a pretty standard 0,1,2 format counting the number of alleles per-genotype. Rows are SNPs, columns are individuals.

Next, we'll run the permutations in parallel for faster processing, storing the optional PCA results for plotting. This step may take a bit of time!

```{r}
perms <- run_permutation(sn, # data
                         n = 100, # 100 permutations
                         par = 6, # in parallel using 6 cores
                         facet = facet, # population info
                         store_pca = TRUE) # save PCA results
```

We can then use PCAssess to plot the results quickly.

```{r}
plot_permutation_res(perms, 
                     n_perm_pcas = 2) # plot two bootstrapped PCAs with real
```

In this case, it's clear that the increase in clustering *is not* more than expected by chance alone, and thus we should use caution if we want to use "highly informative SNPs" for population assignment or management decisions!

# Produce plots for Figure 5

Lastly, we can produce fancier plots if we'd like--this chunk uses `ggplot2` and `cowplot` to produce a pretty plot comparing structuring results with all SNPs to that in the top 5% by $F_{ST}$, including results from the permutation which shows the increase is as-expected by chance.

```{r}
# make the figure parts for publication
pd <- perms$real_pca$all_vars[,.(PC1, PC2, pop)]
pd[,class := "All SNPs"]
pd <- rbind(pd, cbind(perms$real_pca$selected[,.(PC1, PC2, pop)], class = "Top 5% FST"))

library(ggplot2); library(cowplot)
Fig5A <- ggplot(pd, aes(x = PC1, y = PC2, color = pop)) + 
  geom_point(show.legend = FALSE, size = 4) +
  facet_grid(~class) +
  khroma::scale_color_batlow(discrete = TRUE) +
  theme_bw() +
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_text(size = 18),
        strip.text = element_text(size = 18),
        strip.background = element_blank())

Fig5B <- ggplot(perms$null_distribution, aes(x = delta_Fstat)) +
  geom_density() +
  geom_vline(xintercept = perms$observed_values$delta_Fstat, color = "red") +
  geom_label(data = data.frame(x = perms$observed_values$delta_Fstat,
                                        lab = paste0("p = ", perms$pvalues["delta_Fstat"]),
                                        y = max(density(perms$null_distribution$delta_Fstat)$y)*1.1),
                      aes(x = x, y = y, label = lab), hjust = "inward") +
  coord_cartesian(clip = "off") +
  theme_bw() +
  theme(axis.text = element_text(size = 15),
        axis.title = element_text(size = 18)) +
  xlab(bquote(Delta * "F (change in clustering)")) +
  ylab("Density")

Fig5 <- plot_grid(Fig5A, Fig5B, align = "hv", axis = "lr", nrow = 2)
save_plot("../figures/fig5.pdf", Fig5, base_height = 15, base_width = 11)  
  
```

